packages/beacon-node/src/chain/blocks/importBlock.ts

```diff
@@ -221,8 +227,12 @@ export async function importBlock(
 
   // 5. Compute head. If new head, immediately stateCache.setHeadState()
 
+  const {proposerIndex, slot} = block.message;
+  const useProposerHead = this.beaconProposerCache.get(proposerIndex) !== undefined; // If the block is proposed by us, we calculate newHead using GetProposerHead, else GetCanonicialHead
```

Comment:
Would like a second opinion to see if this make sense.

If we receive a block proposed by others, we just call `this.recomputeForkChoiceHead()` without any logic from `getProposerHead()`.

Reply:
I don't see why we need to treat our proposed blocks vs others. We want to always recompute  forkchoice head (the old/default version) to:
- exchange "status" message to other peers
- set head state to state cache
- get duties

plus having "predict (`UpdateHeadOpt.GetPredictedProposerHead`)" option for things already happen doesn't look right.

Reply:
> plus having "predict (`UpdateHeadOpt.GetPredictedProposerHead`)" option for things already happen doesn't look right.

Sorry there was a typo. Was meant to be 
```
onst newHead = useProposerHead
    ? this.recomputeForkChoiceHead(UpdateHeadOpt.GetProposerHead, slot)
    : this.recomputeForkChoiceHead();
```

Reply:
> I don't see why we need to treat our proposed blocks vs others. We want to always recompute forkchoice head (the old/default version) to:
> 
> * exchange "status" message to other peers
> * set head state to state cache
> * get duties

I guess my rationale is that say for example, we are proposing a block not on the canonical head block `B` because it was late (or other reason), but to build on `B`'s parent block `A`. 

This logic was determined during block proposal which it calls `forkchoice.getProposerHead()`. It makes sense during this time that the head cache `forkchoice.head` should be updated from `B` to `A`.

After publishing this block, we `importBlock()` is called. If we call the old/default version of `updateHead()` for `newHead`, it would yield `B`. So here we have `oldHead = A` and `newHead = B`. 

Then `if (newHead.blockRoot !== oldHead.blockRoot)` is triggered, updating the head state of `StateContextCache` to `B`, while `forkchoice.head` is still `A`. I am not sure if this is correct.